### ISSUE TITLE
Add analyzer/code fix for target-typed new expressions

### ISSUE THREAD

**Issue by davkean** [dotnet/roslyn#46700]
**Version Used**: 
Version 16.8.0 Preview 2.0 [30404.163.master] with C# 9.0

It would be great to have analyzer + code fix that detected unnecessary type specifications in a new expression and suggested using the new target-typed new expressions:

**Steps to Reproduce**:

``` C#
private readonly object _lock = new object();

private static DictionaryEqualityComparer<TKey, TValue> Instance { get; } = new DictionaryEqualityComparer<TKey, TValue>();

List<string> values = new List<string>();
```

**Expected Behavior**:

``` C#
private readonly object _lock = new();

private static DictionaryEqualityComparer<TKey, TValue> Instance { get; } = new();

List<string> values = new();
```

**Comment by CyrusNajmabadi** [comment]
This will likely need to match how we support 'var'.  i.e. 'always/never/when-apperant'.

**Comment by sharwell** [comment]
My proposal:

* Always offer this diagnostic where available for fields
* Always offer this diagnostic where available for local variables and the 'use var' configuration for the line does not indicate that implicit types are preferred
* Offer a refactoring where available for local variables when the diagnostic is not reported, but place the refactoring at lower priority than the "Use implicit type" refactoring

**Comment by CyrusNajmabadi** [comment]
> Further manual configuration of this diagnostic is not necessary. Users who do not want this feature in C# 9+ can turn off the diagnostic using the existing Suppress or Configure functionality.

What about users who want a diagnostic to use `new()` for cases like `Foo(new Dictionary<string, int>())`?

**Comment by CyrusNajmabadi** [comment]
> What about users who want a diagnostic to use new() for cases like Foo(new Dictionary<string, int>())?

Design conclusion.  While thsi case is interesting, it is orthogonal to the cases @sharwell covers above.  As such, we can continue with his plan without impacting this.  We will wait to see if there is enough feedback from users that they care about the less-explicit target-typing cases to warrant us supporting a diagnostic/fix to find and update those as well.  Note: if there is user interest, we believe this would be a *different* diagnostic that could be opted into indepedently of the one that covers fields/locals.

**Comment by Youssef1313** [comment]
The analyzer was added in https://github.com/dotnet/roslyn/pull/47996. @CyrusNajmabadi Should this be closed?

**Comment by Youssef1313** [comment]
@CyrusNajmabadi I think this can be closed :)

**Comment by CyrusNajmabadi** [comment]
Yup.  I'm fine closing this for now and waiting for additional feedback if our existing analyzer/fixer is not sufficient.  Thanks!
